File: benchmark-framework-source/src/main/java/com/benchmark/core/BenchmarkResult.java
package com.benchmark.core;

import com.benchmark.metrics.MetricsSnapshot;

import java.util.Collections;
import java.util.List;
import java.util.LongSummaryStatistics;

/**
 * Immutable record of all metrics captured for one (task, concurrencyLevel) pair.
 */
public final class BenchmarkResult {

    private final String            taskName;
    private final int               concurrencyLevel;
    private final long              totalIterations;
    private final long              totalDurationMs;
    private final List<Long>        iterationTimesNs;    // per-iteration latencies
    private final List<MetricsSnapshot> metricsSnapshots;

    // Pre-computed aggregates
    private final double avgCpuPercent;
    private final double peakCpuPercent;
    private final double avgMemoryMB;
    private final double peakMemoryMB;

    public BenchmarkResult(String taskName,
                           int concurrencyLevel,
                           long totalIterations,
                           long totalDurationMs,
                           List<Long> iterationTimesNs,
                           List<MetricsSnapshot> metricsSnapshots,
                           double avgCpuPercent,
                           double peakCpuPercent,
                           double avgMemoryMB,
                           double peakMemoryMB) {
        this.taskName          = taskName;
        this.concurrencyLevel  = concurrencyLevel;
        this.totalIterations   = totalIterations;
        this.totalDurationMs   = totalDurationMs;
        this.iterationTimesNs  = Collections.unmodifiableList(iterationTimesNs);
        this.metricsSnapshots  = Collections.unmodifiableList(metricsSnapshots);
        this.avgCpuPercent     = avgCpuPercent;
        this.peakCpuPercent    = peakCpuPercent;
        this.avgMemoryMB       = avgMemoryMB;
        this.peakMemoryMB      = peakMemoryMB;
    }

    // -------------------------------------------------------------------------
    // Derived metrics

    /** Operations per second across the whole run. */
    public double getThroughput() {
        return totalDurationMs > 0
                ? (totalIterations * 1000.0) / totalDurationMs
                : 0;
    }

    /** Average latency per iteration in milliseconds. */
    public double getAvgLatencyMs() {
        if (iterationTimesNs.isEmpty()) return 0;
        LongSummaryStatistics s = iterationTimesNs.stream()
                .mapToLong(Long::longValue).summaryStatistics();
        return s.getAverage() / 1_000_000.0;
    }

    /** Median latency in ms. */
    public double getMedianLatencyMs() {
        if (iterationTimesNs.isEmpty()) return 0;
        List<Long> sorted = new java.util.ArrayList<>(iterationTimesNs);
        Collections.sort(sorted);
        int mid = sorted.size() / 2;
        return sorted.size() % 2 == 0
                ? (sorted.get(mid - 1) + sorted.get(mid)) / 2_000_000.0
                : sorted.get(mid) / 1_000_000.0;
    }

    /** 95th-percentile latency in ms. */
    public double getP95LatencyMs() { return percentileMs(95); }

    /** 99th-percentile latency in ms. */
    public double getP99LatencyMs() { return percentileMs(99); }

    private double percentileMs(int p) {
        if (iterationTimesNs.isEmpty()) return 0;
        List<Long> sorted = new java.util.ArrayList<>(iterationTimesNs);
        Collections.sort(sorted);
        int idx = (int) Math.ceil(p / 100.0 * sorted.size()) - 1;
        return sorted.get(Math.max(0, Math.min(idx, sorted.size() - 1))) / 1_000_000.0;
    }

    // -------------------------------------------------------------------------
    // Accessors

    public String               getTaskName()         { return taskName; }
    public int                  getConcurrencyLevel() { return concurrencyLevel; }
    public long                 getTotalIterations()  { return totalIterations; }
    public long                 getTotalDurationMs()  { return totalDurationMs; }
    public List<Long>           getIterationTimesNs() { return iterationTimesNs; }
    public List<MetricsSnapshot> getMetricsSnapshots() { return metricsSnapshots; }
    public double               getAvgCpuPercent()   { return avgCpuPercent; }
    public double               getPeakCpuPercent()  { return peakCpuPercent; }
    public double               getAvgMemoryMB()     { return avgMemoryMB; }
    public double               getPeakMemoryMB()    { return peakMemoryMB; }

    @Override
    public String toString() {
        return String.format(
            "[%s | concurrency=%-3d] throughput=%,.1f ops/s  avgLatency=%6.2f ms  " +
            "p99=%6.2f ms  avgCPU=%5.1f%%  peakMem=%6.1f MB",
            taskName, concurrencyLevel,
            getThroughput(), getAvgLatencyMs(), getP99LatencyMs(),
            avgCpuPercent, peakMemoryMB
        );
    }
}

File: python/benchmark/core/benchmark_result.py
from __future__ import annotations

import math
from dataclasses import dataclass

from benchmark.metrics.metrics_snapshot import MetricsSnapshot


@dataclass(frozen=True)
class BenchmarkResult:
    """Immutable record of metrics for one (task, concurrency_level) pair."""

    task_name: str
    concurrency_level: int
    total_iterations: int
    total_duration_ms: int
    iteration_times_ns: tuple[int, ...]
    metrics_snapshots: tuple[MetricsSnapshot, ...]
    avg_cpu_percent: float
    peak_cpu_percent: float
    avg_memory_mb: float
    peak_memory_mb: float

    def get_throughput(self) -> float:
        if self.total_duration_ms <= 0:
            return 0.0
        return (self.total_iterations * 1000.0) / self.total_duration_ms

    def get_avg_latency_ms(self) -> float:
        if not self.iteration_times_ns:
            return 0.0
        return (sum(self.iteration_times_ns) / len(self.iteration_times_ns)) / 1_000_000.0

    def get_median_latency_ms(self) -> float:
        if not self.iteration_times_ns:
            return 0.0
        sorted_ns = sorted(self.iteration_times_ns)
        n = len(sorted_ns)
        mid = n // 2
        if n % 2 == 0:
            return (sorted_ns[mid - 1] + sorted_ns[mid]) / 2_000_000.0
        return sorted_ns[mid] / 1_000_000.0

    def get_p95_latency_ms(self) -> float:
        return self._percentile_ms(95)

    def get_p99_latency_ms(self) -> float:
        return self._percentile_ms(99)

    def _percentile_ms(self, p: int) -> float:
        if not self.iteration_times_ns:
            return 0.0
        sorted_ns = sorted(self.iteration_times_ns)
        idx = int(math.ceil(p / 100.0 * len(sorted_ns))) - 1
        idx = max(0, min(idx, len(sorted_ns) - 1))
        return sorted_ns[idx] / 1_000_000.0

    def __str__(self) -> str:
        return (
            f"[{self.task_name} | concurrency={self.concurrency_level:<3}] "
            f"throughput={self.get_throughput():,.1f} ops/s  "
            f"avgLatency={self.get_avg_latency_ms():6.2f} ms  "
            f"p99={self.get_p99_latency_ms():6.2f} ms  "
            f"avgCPU={self.avg_cpu_percent:5.1f}%  "
            f"peakMem={self.peak_memory_mb:6.1f} MB"
        )
